common/src/main/java/org/apache/sedona/common/S2Geography/WKBWriter.java

```diff
@@ -0,0 +1,378 @@
+   * @return the byte array containing the WKB
+   */
+  public byte[] write(S2Geography geog) {
+    try {
+      byteArrayOS.reset();
+      write(geog, byteArrayOutStream);
+    } catch (IOException ex) {
+      throw new RuntimeException("Unexpected IO exception: " + ex.getMessage());
+    }
+    return byteArrayOS.toByteArray();
+  }
+
+  /**
+   * Writes a {@link S2Geography} to an {@link OutStream}.
+   *
+   * @param geog the geometry to write
+   * @param os the out stream to write to
+   * @throws IOException if an I/O error occurs
+   */
+  public void write(S2Geography geog, OutStream os) throws IOException {
+    int dim = geog.dimension(); //  0=point, 1=linestring, 2=polygon, -1=mixed/empty
+    int count = geog.numShapes(); //  1=single, >1=multi
+
+    String s =
+        "Cannot write geometry with dimension="
+            + dim
+            + " and shape count="
+            + count
+            + " (type: "
+            + geog.getClass().getName()
+            + ")";
+
+    // 1) Mixed or empty collection
+    if (dim == -1) {
+      if (geog instanceof GeographyCollection) {
+        writeGeographyCollection(
+            WKBConstants.wkbGeometryCollection, (GeographyCollection) geog, os);
+        return;
+      }
+      throw new IllegalArgumentException(s);
+    }
+
+    // 2) Dispatch on the real dimensions
+    switch (dim) {
+      case 0: // POINT or MULTI_POINT
+        if (count == 1) {
+          // exactly one point
+          writePoint(WKBConstants.wkbPoint, (SinglePointGeography) geog, os);
+        } else {
+          // more than one → treat as multipoint
+          writePoint(WKBConstants.wkbMultiPoint, (PointGeography) geog, os);
+        }
+        return;
+
+      case 1: // LINESTRING or MULTI_LINESTRING
+        if (count == 1) {
+          writePolyline(WKBConstants.wkbLineString, (SinglePolylineGeography) geog, os);
+        } else {
+          writePolyline(WKBConstants.wkbMultiLineString, (PolylineGeography) geog, os);
+        }
+        return;
+
+      case 2: // POLYGON or MULTI_POLYGON
+        if (count == 1) {
+          writePolygon(WKBConstants.wkbPolygon, (PolygonGeography) geog, os);
+        } else {
+          writeGeographyCollection(WKBConstants.wkbMultiPolygon, (MultiPolygonGeography) geog, os);
+        }
+        return;
+
+      default:
+        // Should never happen once dim ∈ {0,1,2,-1} is enforced
+        throw new IllegalArgumentException(s);
+    }
+  }
+
+  private void writePoint(int geometryType, PointGeography pt, OutStream os) throws IOException {
+    writeByteOrder(os);
+    writeGeometryType(geometryType, pt, os);
+    if (pt.numShapes() == 0) {
+      writeNaNs(outputDimension, os);
+    } else {
+      S2Point p = pt.getPoints().get(0);
+      S2LatLng ll = new S2LatLng(p);
+
+      // round to 6 decimal places //TODO: 16 decimal
+      double lon = Math.rint(ll.lngDegrees() * 1e16) / 1e16;
+      double lat = Math.rint(ll.latDegrees() * 1e16) / 1e16;
```

Comment:
The WKBWriter does not have take care about precision model. I believe that the handling of maximum significant digits only happens in WKT writer?